cpp/src/io/json/nested_json_gpu.cu

```diff
@@ -0,0 +1,638 @@
+}  // namespace to_stack_op
+
+//------------------------------------------------------------------------------
+// JSON TOKENIZER PUSHDOWN AUTOMATON
+//------------------------------------------------------------------------------
+namespace tokenizer_pda {
+
+// Type used to represent the target state in the transition table
+using StateT = char;
+
+/**
+ * @brief Symbol groups for the input alphabet for the pushdown automaton
+ */
+enum SGID : PdaSymbolGroupIdT {
+  /// Opening brace
+  OBC,
+  /// Opening bracket
+  OBT,
+  /// Closing brace
+  CBC,
+  /// Closing bracket
+  CBT,
+  /// Quote
+  QTE,
+  /// Escape
+  ESC,
+  /// Comma
+  CMA,
+  /// Colon
+  CLN,
+  /// Whitespace
+  WSP,
+  /// Other (any input symbol not assigned to one of the above symbol groups)
+  OTR,
+  /// Total number of symbol groups amongst which to differentiate
+  NUM_PDA_INPUT_SGS
+};
+
+/**
+ * @brief Symbols in the stack alphabet
+ */
+enum STACK_SGID : PdaStackSymbolGroupIdT {
+  /// Symbol representing the JSON-root (i.e., we're at nesting level '0')
+  STACK_ROOT = 0,
+
+  /// Symbol representing that we're currently within a list object
+  STACK_LIST = 1,
+
+  /// Symbol representing that we're currently within a struct object
+  STACK_STRUCT = 2,
+
+  /// Total number of symbols in the stack alphabet
+  NUM_STACK_SGS
+};
+
+/// Total number of symbol groups to differentiate amongst (stack alphabet * input alphabet)
+constexpr PdaSymbolGroupIdT NUM_PDA_SGIDS = NUM_PDA_INPUT_SGS * NUM_STACK_SGS;
+
+/// Mapping a input symbol to the symbol group id
+static __constant__ PdaSymbolGroupIdT tos_sg_to_pda_sgid[] = {
+  OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, WSP, WSP, OTR, OTR, WSP, OTR, OTR, OTR, OTR, OTR,
+  OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, WSP, OTR, QTE, OTR, OTR, OTR,
+  OTR, OTR, OTR, OTR, OTR, OTR, CMA, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR,
+  OTR, CLN, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR,
+  OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OBT, ESC, CBT, OTR,
+  OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR,
+  OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OBC, OTR, CBC, OTR};
+
+/**
+ * @brief Maps a (top-of-stack symbol, input symbol)-pair to a symbol group id of the deterministic
+ * visibly pushdown automaton (DVPA)
+ */
+struct PdaSymbolToSymbolGroupId {
+  template <typename SymbolT, typename StackSymbolT>
+  __device__ __forceinline__ PdaSymbolGroupIdT
+  operator()(thrust::tuple<SymbolT, StackSymbolT> symbol_pair)
+  {
+    // The symbol read from the input
+    auto symbol = thrust::get<0>(symbol_pair);
+
+    // The stack symbol (i.e., what is on top of the stack at the time the input symbol was read)
+    // I.e., whether we're reading in something within a struct, a list, or the JSON root
+    auto stack_symbol = thrust::get<1>(symbol_pair);
```

Comment:
Nit: `auto [symbol, stack_symbol] = symbol_pair`

Reply:
I believe that `thrust::tuple` does not support structured bindings? This comment seems to support this(?):
https://github.com/rapidsai/cudf/pull/9931#discussion_r779019486

Reply:
Ah, `std::tuple` works just fine but not `thrust::tuple`, good to know.